ostd/src/logger.rs

```diff
@@ -10,96 +10,74 @@
 //!
 //! IRQs are disabled while printing. So do not print long log messages.
 
+use core::str::FromStr;
+
 use log::{LevelFilter, Metadata, Record};
+use spin::Once;
+
+static LOGGER: Logger = Logger::new();
 
-use crate::{
-    boot::{kcmdline::ModuleArg, kernel_cmdline},
-    timer::Jiffies,
-};
+struct Logger {
+    backend: Once<&'static dyn log::Log>,
+}
 
-const LOGGER: Logger = Logger {};
+impl Logger {
+    const fn new() -> Self {
+        Self {
+            backend: Once::new(),
+        }
+    }
+}
 
-struct Logger {}
+/// Sets the backend for the global logger, allowing upper-level users to register
+/// their own implemented loggers.
+///
+/// This method only allows setting once; subsequent attempts to set it again will have no effect.
+pub fn set_logger(new_logger: &'static dyn log::Log) {
+    LOGGER.backend.call_once(|| new_logger);
+}
 
 impl log::Log for Logger {
     fn enabled(&self, metadata: &Metadata) -> bool {
+        if let Some(logger) = self.backend.get() {
+            return logger.enabled(metadata);
+        };
+
+        // Default instruction.
         metadata.level() <= log::max_level()
```

Comment:
*This is actually a comment on the original logger implementation. I admit that the problem is not introduced by this PR, but it is just slightly better if it can be addressed together.*

There is no point in checking the log level in `Log::enabled`. The logging macro ([`log_enabled!`](https://docs.rs/log/0.4.22/src/log/macros.rs.html#215-225) and [`log!`](https://docs.rs/log/0.4.22/src/log/macros.rs.html#31-60)) has already checked the log level. This comment also applies to `DefaultLogger::enabled`.